[READY] Add debugging instructions and tools

puremourning opened the pull request:
Adds some stuff that developers might find useful. I'm totally aware that these changes are probably most useful to me personally, so as usual I won't be offended if you don't want to merge them. Patch contains:
 - instructions on debugging python and the python c++ module (it took me a *long* time to do this successfully)
 - `--enable-debug` flag to the build
 - `--dump-path` flag to `run_tests.py`. This is useful if you want to run `nosetests` directly (e.g. as `env PYTHONPATH=`./run_tests.py --dump-path` pdb nosetests ...`
- Parallels provider to the Vagrantfile (as I seem to have a moral objection to using virtual box when spend hard earned on Parallels :)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/454)
<!-- Reviewable:end -->

Valloric commented:
I welcome all of these changes. :)

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with minor comments.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[DEV_SETUP.md, line 27 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KExzf461wmKjE-SB7q4:-KExzf461wmKjE-SB7q5:1213752324) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L27)):*
The heading here is a bit weird. :)

---

*[DEV_SETUP.md, line 29 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KExzkREHDdjLDQFoO_0:-KExzkREHDdjLDQFoO_1:1968975380) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L29)):*
Using singular reads weird in technical documentation. Could you make this less personal?

---

*[DEV_SETUP.md, line 123 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KEy-EtiPYDzOu3trOlI:-KEy-EtiPYDzOu3trOlJ:-1489698146) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L123)):*
First comma should be a colon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KEy-PJQBcOuFk9Rm2bX:-1145861998)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Updated. Thanks for the review.

---

Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[DEV_SETUP.md, line 27 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KExzf461wmKjE-SB7q4:-KF_LBBLgFjG3CrwovEU:-1291034683) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L27)):*
I R DEBUG ALL THE PYTHONS. Changed it to "python layer" which I think is what I was getting at (as different from the c++ layer)

---

*[DEV_SETUP.md, line 29 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KExzkREHDdjLDQFoO_0:-KF_LmKognxjBQ_OAmQK:-1958564428) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L29)):*
Sure, done. I was trying to cover myself - I don't claim to be an expert in this field :). I worry about writing things which sound authoritative, but are not written by an authority!

---

*[DEV_SETUP.md, line 123 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KEy-EtiPYDzOu3trOlI:-KF_MoSQb0mGHmUK_6Vs:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/DEV_SETUP.md#L123)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KF_NcaRkV7_gFsxN7LB:2057389377)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> as before. You should probably update the title to READY now.

@micbou @vheon second review?

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KFbPbgFEGTwQDoEy_tq:-803577411)*
<!-- Sent from Reviewable.io -->

micbou commented:
I need to try these instructions to debug the ycm_core library on Windows. I suspect it will not be that easy.

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[build.py, line 256 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDIpBqEN7Qv9gyy-TG:-KGDIpBrZ2gP2eQ4FMmX:-1792670900) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/build.py#L256)):*
It's better to say `ycm_core library` than `ycm_core.so` because the `.so` is platform specific.

---

*[DEV_SETUP.md, line 72 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDHuXyRwfvGHg732s0:-KGDHuXyRwfvGHg732s1:-858271949) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L72)):*
We could add `sh` to the blocks with shell code for syntax highlighting.

---

*[DEV_SETUP.md, line 102 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKMeOjNJvfiPlUpFU:-KGDKMePgTV399FYKJIK:1429935550) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L102)):*
Same comment as above: `Build the ycm_core library...`

---

*[DEV_SETUP.md, line 107 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKbG1L4859oDacs4A:-KGDKbG2-bRaM_92SNql:-1034959140) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L107)):*
`build.py` instead of `build.sh`.

---

*[run_tests.py, line 216 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KEwG7EpwnMPKYmDAVjn:-KEwG7EqlNuQAANaIUpL:-837656648) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/run_tests.py#L216)):*
We should use `sys.exit` instead of `exit`. [`exit` is for the interactive interpreter shell](https://docs.python.org/2/library/constants.html?highlight=exit#constants-added-by-the-site-module).

---

*[run_tests.py, line 217 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKwjZGc1Yp4VrORRF:-KGDKwjZGc1Yp4VrORRG:-350437375) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/run_tests.py#L217)):*
Unneeded blank line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KGDIdAztDswb6cfyWQa:976790667)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[build.py, line 256 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDIpBqEN7Qv9gyy-TG:-KHFq_DxcWF-rTXzHXo6:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/build.py#L256)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> It's better to say `ycm_core library` than `ycm_core.so` because the `.so` is platform specific.</details>
Done.

---

*[DEV_SETUP.md, line 72 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDHuXyRwfvGHg732s0:-KHFqlGhCAqMnYwrIIWT:179868919) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L72)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We could add `sh` to the blocks with shell code for syntax highlighting.</details>
/ycmnDone.

---

*[DEV_SETUP.md, line 102 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKMeOjNJvfiPlUpFU:-KHFqr1TjbJVaqrEHVXJ:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L102)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Same comment as above: `Build the ycm_core library...`</details>
Done.

---

*[DEV_SETUP.md, line 107 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKbG1L4859oDacs4A:-KHFquLEC_girT_tj_tE:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/DEV_SETUP.md#L107)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> `build.py` instead of `build.sh`.</details>
Done.

---

*[run_tests.py, line 216 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KEwG7EpwnMPKYmDAVjn:-KHFr1iCTnrAyRJBslMY:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/c4a370fb16c8e69f6594bd4b00a6a16ec7931f21/run_tests.py#L216)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> We should use `sys.exit` instead of `exit`. [`exit` is for the interactive interpreter shell](https://docs.python.org/2/library/constants.html?highlight=exit#constants-added-by-the-site-module).</details>
Done.

---

*[run_tests.py, line 217 \[r2\]](https://reviewable.io:443/reviews/valloric/ycmd/454#-KGDKwjZGc1Yp4VrORRF:-KHFr1UMekcnU1gR6ndO:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/66f944f5da728447f58ff84729bc9caa29ff9a6d/run_tests.py#L217)):*
<details><summary><i>Previously, micbou wrote…</i></summary>
> Unneeded blank line.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454)*
<!-- Sent from Reviewable.io -->

homu commented:
:umbrella: The latest upstream changes (presumably #481) made this pull request unmergeable. Please resolve the merge conflicts.

puremourning commented:
Fixed the merge conflict

<details><summary><i>Previously, homu (Homu) wrote…</i></summary>
> :umbrella: The latest upstream changes (presumably #481) made this pull request unmergeable. Please resolve the merge conflicts.</details>


---

Review status: 1 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KHRZukVTBWN2DVc_e94:-1948324710)*
<!-- Sent from Reviewable.io -->

micbou commented:
Let's merge this. If we find an easy way to debug the ycmd library on Windows (i.e. without Visual Studio), we'll update the instructions.

@homu r+

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6126849/badge)](https://coveralls.io/builds/6126849)
> 
> Coverage remained the same at 84.957% when pulling **39996c5237a119cb87880af49280e488d7a9a364 on puremourning:add-debugging-tools** into **b74add9c32f6edd71e165a7151fac12cf8a9a32c on Valloric:master**.
> </details>


---

Reviewed 3 of 3 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/454#-:-KHG3miUTVKk6i1qZ6mK:-596049094)*
<!-- Sent from Reviewable.io -->

homu commented:
:pushpin: Commit 39996c5 has been approved by `micbou`

<!-- @homu r=micbou 39996c5237a119cb87880af49280e488d7a9a364 -->

homu commented:
:zap: Test exempted - [status](https://travis-ci.org/Valloric/ycmd/builds/129269297)
